### PR TITLE
feat/add support for timeout 

### PR DESCRIPTION
**Pull request by 0xArdi**
## Proposed changes

Adds support for arbitrary functions as tasks with timeouts.  

## Types of changes

What types of changes does your code introduce? (A **breaking change** is a fix or feature that would cause existing functionality and APIs to not work as expected.)
_Put an `x` in the box that applies_

- [ ] Non-breaking fix (non-breaking change which fixes an issue)
- [ ] Breaking fix (breaking change which fixes an issue)
- [x] Non-breaking feature (non-breaking change which adds functionality)
- [ ] Breaking feature (breaking change which adds functionality)
- [ ] Refactor (non-breaking change which changes implementation)
- [ ] Messy (mixture of the above - requires explanation!)

## Checklist

_Put an `x` in the boxes that apply._

- [x] I have read the [CONTRIBUTING](../blob/main/CONTRIBUTING.md) doc
- [x] I am making a pull request against the `main` branch (left side). Also you should start your branch off our `main`.
- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have locally run services that could be impacted and they do not present failures derived from my changes

### CODE REVIEW

File: packages/valory/connections/ledger/base.py [hunk 1]
```diff
@@ -92,13 +94,49 @@ async def run_async(
...
+        # we run the passed function using the default executor
+        running_func = self.loop.run_in_executor(self.executor, func)
+        try:
+            # func_result will carry the
+            # value the function returns
+            func_result = await asyncio.wait_for(running_func, timeout=timeout)
+            return func_result
+        except TimeoutError:
+            self.logger.warning(f"Task could not finish in {timeout}s.")
+            return None
```
Comment (0xArdi) [inline packages/valory/connections/ledger/base.py:138]: Better error messages required.

File: packages/valory/connections/ledger/base.py [hunk 2]
```diff
@@ -92,13 +94,49 @@ async def run_async(
         :return: the return value of the function.
         """
         try:
-            response = await self.loop.run_in_executor(
-                self.executor, func, api, message, dialogue
-            )
+            if inspect.iscoroutinefunction(func):
+                # If it is a coroutine, no need to run it in an executor
+                # This can happen if the handler is async.
+                task = func(api, message, dialogue)  # type: ignore
```
Comment (0xArdi) [inline packages/valory/connections/ledger/base.py:100]: This ensures compatibility with coroutines and non-coroutines.

File: tests/test_connections/test_ledger/test_ledger.py [hunk 3]
```diff
@@ -135,6 +138,61 @@ async def test_not_hanging(self) -> None:
...
+    async def dummy_async_func(self, sleep: float) -> bool:
+        """A dummy async function that sleeps and returns True."""
+        await asyncio.sleep(sleep)
+        return True
+
+    @pytest.mark.asyncio
+    async def test_wait_for_happy_path(self) -> None:
+        """Tests that wait_for works when timeout is bigger than execution time of callable."""
+        should_finish_in = 0.5  # 0.5 seconds
+        timeout = 1  # we give 2 the amount of time for it to finish
```
Comment (Adamantios) [inline tests/test_connections/test_ledger/test_ledger.py:170]: Maybe this is more self-explainable?

```suggestion
        should_finish_in = 0.5
        tolerance = 0.1
        timeout = should_finish_in + tolerance
```

File: tests/test_connections/test_ledger/test_ledger.py [hunk 4]
```diff
@@ -135,6 +138,61 @@ async def test_not_hanging(self) -> None:
...
+        return_value = await self.dispatcher.wait_for(
+            lambda: self.dummy_func(should_finish_in), timeout=timeout
+        )
+        assert return_value, "dummy_func() should've returned True"
+
+    @pytest.mark.asyncio
+    async def test_wait_for_raise_excp(self) -> None:
+        """Tests that wait_for works when timeout is less than execution time of callable."""
+        should_finish_in = 0.5  # 0.5 seconds
+        timeout = 0.25  # we give 2 the amount of time for it to finish
```
Comment (Adamantios) [inline tests/test_connections/test_ledger/test_ledger.py:181]: And how about something like this?

```suggestion
        # we increase the timeout by a tiny amount to test that a `TimeoutError` will be raised
        timeout = 0.25
        timeout_increase = 0.1
        should_finish_in = timeout + timeout_increase
```

**Review by Adamantios — COMMENTED** [review]
Looks good, only some minor comments. However, I would suggest that we change the base to `test/node-blocking`.

File: packages/valory/connections/ledger/ledger_dispatcher.py [hunk 5]
```diff
@@ -285,8 +291,12 @@ def get_transaction_receipt(
             and self.connection_state.get() == ConnectionStates.connected
         ):
             try:
-                transaction = api.get_transaction(
-                    message.transaction_digest.body, raise_on_try=True
+                transaction = await self.wait_for(
```
Comment (DavidMinarsch) [inline packages/valory/connections/ledger/ledger_dispatcher.py:294]: why the change, the `get_transaction` remained unchanged.

Comment (0xArdi) [reply packages/valory/connections/ledger/ledger_dispatcher.py:294]: The method is async, get_transaction is sync. It can block the execution.

**Comment by 0xArdi** [comment]
> Looks good, only some minor comments. However, I would suggest that we change the base to `test/node-blocking`.

@Adamantios given that this is a feat, and `test/node-blocking` is a test. Should we maybe rebase `test/node-blocking` to this branch. Then we can merge them both under `feat/add-timeout-support` on main?

**Comment by Adamantios** [comment]
> > Looks good, only some minor comments. However, I would suggest that we change the base to `test/node-blocking`.
> 
> @Adamantios given that this is a feat, and `test/node-blocking` is a test. Should we maybe rebase `test/node-blocking` to this branch. Then we can merge them both under `feat/add-timeout-support` on main?

Yes, this would be much better! I will do that.

**Comment by 0xArdi** [comment]
@Adamantios yes, its a [`TimeoutError`](https://github.com/valory-xyz/open-autonomy/runs/8269101248?check_suite_focus=true#step:8:6686)

File: .github/workflows/main_workflow.yml [hunk 6]
```diff
@@ -164,7 +164,7 @@ jobs:
         os: [ ubuntu-latest, macos-latest, windows-latest ]
         python-version: [ "3.7", "3.8", "3.9", "3.10" ]
 
-    timeout-minutes: 50
+    timeout-minutes: 55
```
Comment (Adamantios) [inline .github/workflows/main_workflow.yml:167]: Tests require more time now (~2min).

File: tests/test_connections/test_ledger/test_ledger_api.py [hunk 7]
```diff
@@ -480,11 +480,13 @@ async def test_no_raw_tx(
         ("get_transaction_receipt", "is_transaction_settled", "get_transaction"),
     )
     @pytest.mark.parametrize("retries", (0, 5, 20))
+    @pytest.mark.parametrize("retry_timeout", (0.1,))
```
Comment (Adamantios) [inline tests/test_connections/test_ledger/test_ledger_api.py:483]: https://github.com/valory-xyz/open-autonomy/pull/1295#issuecomment-1241982698

**Review by DavidMinarsch — APPROVED** [review]
